helix-term/src/handlers/signature_help.rs

```diff
@@ -0,0 +1,345 @@
+use helix_event::{
+    cancelable_future, cancelation, register_hook, send_blocking, CancelRx, CancelTx,
+};
+use helix_lsp::lsp;
+use helix_view::document::Mode;
+use helix_view::events::{DocumentDidChange, SelectionDidChange};
+use helix_view::handlers::lsp::{SignatureHelpEvent, SignatureHelpInvoked};
+use helix_view::Editor;
+use tokio::sync::mpsc::Sender;
+use tokio::time::Instant;
+
+use crate::commands::Open;
+use crate::compositor::Compositor;
+use crate::events::{OnModeSwitch, PostInsertChar};
+use crate::handlers::{rope_ends_with, Handlers};
+use crate::ui::lsp::SignatureHelp;
+use crate::ui::Popup;
+use crate::{job, ui};
+
+#[derive(Debug)]
+enum State {
+    Open,
+    Closed,
+    Pending { request: CancelTx },
+}
+
+/// debounce timeout in ms, value taken from VSCode
+/// TODO: make this configurable?
+const TIMEOUT: u64 = 120;
+
+#[derive(Debug)]
+pub(super) struct SignatureHelpHandler {
+    trigger: Option<SignatureHelpInvoked>,
+    state: State,
+}
+
+impl SignatureHelpHandler {
+    pub fn new() -> SignatureHelpHandler {
+        SignatureHelpHandler {
+            trigger: None,
+            state: State::Closed,
+        }
+    }
+}
+
+impl helix_event::AsyncHook for SignatureHelpHandler {
+    type Event = SignatureHelpEvent;
+
+    fn handle_event(
+        &mut self,
+        event: Self::Event,
+        timeout: Option<tokio::time::Instant>,
+    ) -> Option<Instant> {
+        match event {
+            SignatureHelpEvent::Invoked => {
+                self.trigger = Some(SignatureHelpInvoked::Manual);
+                self.state = State::Closed;
+                self.finish_debounce();
+                None
+            }
+            SignatureHelpEvent::Trigger => {
+                if self.trigger.is_none() {
+                    self.trigger = Some(SignatureHelpInvoked::Automatic)
+                }
+                Some(Instant::now() + Duration::from_millis(TIMEOUT))
+            }
+            SignatureHelpEvent::ReTrigger => {
+                if matches!(self.state, State::Closed) {
+                    timeout
+                } else {
+                    if self.trigger.is_none() {
+                        self.trigger = Some(SignatureHelpInvoked::Automatic)
+                    }
+                    Some(Instant::now() + Duration::from_millis(TIMEOUT))
+                }
+            }
+            SignatureHelpEvent::Cancel => {
+                if !matches!(self.state, State::Closed) {
+                    self.state = State::Closed;
+                }
+                None
+            }
+            SignatureHelpEvent::RequestComplete { open: opened } => {
+                if opened {
+                    // don't cancel rerequest that were already triggered
+                    if let State::Pending { request } = &self.state {
+                        if !request.is_closed() {
+                            return timeout;
+                        }
+                    }
```

Comment:
basically if request A fininshes about the same time as the last timeout finishes (so the state is again pending). Usually that would cancel the old request but if the old request was already done the cancelation doesn't work. I that case we want to avoid canceling the newer request here. Its a bit of an edgecase but I did run into this while testing something

Reply:
Ahhh ok, that makes sense. Could we add a comment here that explains this edge case?